### PR TITLE
Fix: #4856 #4857

### PR DESCRIPTION
**Pull request by leberus**
Should fix issues #4856 and #4857

### CODE REVIEW

File: libr/debug/p/native/linux/linux_debug.c [hunk 1]
```diff
@@ -195,10 +214,27 @@ bool linux_set_options (RDebug *dbg, int pid) {
...
 int linux_attach (RDebug *dbg, int pid) {
 	linux_set_options (dbg, pid);
-	int ret = ptrace (PTRACE_ATTACH, pid, 0, 0);
+	int ret = ptrace (PTRACE_ATTACH, pid, NULL, NULL);
 	if (ret != -1) {
 		perror ("ptrace (PT_ATTACH)");
 	}
+	
+	if (dbg->h) {
+		RList *list = dbg->h->threads (dbg, pid);
```
Comment (alvarofe) [inline libr/debug/p/native/linux/linux_debug.c]: don't use dbg->h->threads since you are already in the linux backend, instead use linux_thread_list

**Comment by alvarofe** [comment]
Do we have a test for this?

**Comment by leberus** [comment]
Right now Im afk, but its quite easy to test it. just write a small program
that created a thread, then try to attach r2 to the main process and you'll
see that just the main process is stopped while the other thread is still
running. With this patch the main process together with all threads are
stopped (and then continued). once I get at home I can upload a PoC

El dia 24/12/2016 18:58, "Álvaro Felipe Melchor" <notifications@github.com>
va escriure:

**Comment by leberus** [comment]
I'll try to record an asciicinema

El dia 24/12/2016 7:11 PM, "Oscar Salvador" <osalvador.vilardaga@gmail.com>
va escriure:

**Comment by alvarofe** [comment]
I know that is easy to test that but did you test to stop, continue one of threads, put breakpoints and so on so forth?

**Comment by leberus** [comment]
I did some tests, but i'll test further. I'll come back once done

El dia 24/12/2016 7:16 PM, "Álvaro Felipe Melchor" <notifications@github.com>
va escriure:

**Comment by leberus** [comment]
I've seen one thing:

In **cmd_debug_continue**, either you call "dc" or "dc X", you reach the same point: **r_debug_continue**.
This function only takes one parameter, so it doesn't really know if it must continue all threads or just one.
With r2 current code, it only continues one thread (even if you specify "dc").

What about adding a bool parameter to the function, like **cont_all_threads**. So there is a real difference between "dc" and "dc X". If not a function, what about a field in RDebug structure?

Thanks

**Comment by radare** [comment]
Yep do this change

**Comment by radare** [comment]
pls rebase

**Comment by leberus** [comment]
I'll do, but later, im still celebrating christmas with family.

El dia 26/12/2016 2:27 PM, "radare" <notifications@github.com> va escriure:

**Comment by radare** [comment]
Yeah everyone is busy irl these days xD

**Comment by leberus** [comment]
I did some tests and so far is working. But there are still things that must be improved/fixed.
For instance, if the main thread waits for a thread to be done (with pthread_join):
If we want to continue all threads, what we do right now is PTRACE_CONT and then we retrieve the status with: 
`waitpid (pid, &status, __WALL)`
But his call is gonna block since main_thread is waiting one of its childs to finish before it continues, so I had to add the flag **WNOHANG**.
There are a couple of things more, but unfortunately till 3th (when I'm back at the office) I can't test/fix further.

**Comment by radare** [comment]
the WNOHANG flag introduces regressions on osx and linux backends for simple hello world programs that block on a syscall like read() it is a no -way. unless you manage to call the waitpid without this after sending the cont to all the threads. i would go for using the WNOHANG for all threads and then calling waitpid to the selected thread without this flag.

pls send a pr updating this and test it. we need more tests for the debugger

**Comment by leberus** [comment]
Ok, i'll send a pr once i get it working. I start tomorrow again ;)

El dia 31/12/2016 19:12, "radare" <notifications@github.com> va escriure:

**Comment by radare** [comment]
👍

**Comment by leberus** [comment]
I'm preparing a "big" refactoring of linux_debug.c (for the threads stuff and so son), but just wanted to say that I'll do the PR tomorrow.

**Comment by radare** [comment]
awesome :)  I <3 refactors

**Pull request merged**